UI: Add ability to make screenshots

cg2121 opened the pull request:
### Description
This adds the ability to take screenshots of the preview with a hotkey. It currently only saves to png and doesn't have any settings, so it saves to where the recordings are saved.

### Motivation and Context
https://ideas.obsproject.com/posts/4/add-a-screenshot-button

This is simpler than #1269 because it uses QImage to save the screenshot.

The code is based off of @Palakis screenshot code in the websockets plugin.

### How Has This Been Tested?
Took screenshots and made sure they saved correctly.

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

RytoEX commented:
> Not quite sure how to deal with the libpng dependency on other OSs.

We already use libpng as part of the prebuilt deps on Windows and macOS.  That is reflected on the [Release Dependency Versions wiki page](https://github.com/obsproject/obs-studio/wiki/Release-Dependency-Versions).

Personally, I still feel like this sort of thing is better off as an optional/third-party plugin that isn't part of the core set of plugins.

cg2121 commented:
It now saves with full size output instead of the scaled version. I removed the plugin and now use QImage to save the screenshot.

It is based off of @Palakis code in the websocket plugin. https://github.com/Palakis/obs-websocket/commit/0b9366fa28bfd6f5ece53bbd900aba656da215e5

WizardCM commented:
Awesome! Tested it on Windows 10, works like a dream.

It currently supports full transparency on the output. Personally I think that's awesome and could allow for some really cool uses, however I can also think of reasons why people would want to have the screenshot include the black background. Maybe a setting under General?

cg2121 commented:
Should screenshots have own settings section? I can think of a few settings it might have:

- png or jpg (maybe others as well)
- save path
- allow transparency (if image format supports it)
- filename prefix
- filename suffix

Or would this be overkill?

Fenrirthviti commented:
I feel keeping this as simple as possible is the right move. I would even argue allowing for transparency in the screenshot, while potentially useful in some extremely niche cases, would go against expectations.

I'd vote in favor of keeping the background black and not allowing for transparency. The simpler it is, the better.

cg2121 commented:
I've changed it to not have transparency.

I've also added the ability to take screenshot of specific source. This is only exposed to the frontend api.

WizardCM commented:
I agree with keeping it as simple as possible, and don't see the need for a bunch of settings.

WizardCM commented:
I can confirm the new version doesn't have transparency.

New issue: hitting the hotkey freezes the UI for a split second. If you hit the screenshot hotkey a few times (say, 3-4 times a second for 3 seconds) the UI basically hangs for a few seconds on top.

cg2121 commented:
With the latest update, does the UI still freeze for you @WizardCM?

WizardCM commented:
@cg2121 Yep, still freezes. Not sure if it matters, but this is on Windows 10 with Qt 5.14.2 specifically.

jp9000 commented:
I fixed the lag issues in the UI, and deferred it to the graphics thread.

I made it so that there's a single frame delay between each step in the screenshot process: render the texture, wait a frame, download texture, wait a frame, lock RGBA data and copy it to the QImage.  After that, muxing of the QImage to PNG file is now performed in a separate thread entirely, and when that is complete, the screenshot data automatically frees itself.

Doing these things ensures that there is no significant lag either in the graphics thread or in the main UI thread.

WizardCM commented:
Just tested: UI no longer lags, and all the new features mentioned by Jim work great. :D

GXTX commented:
Although it seems you all are set on this I'll give my input as a user RE: transparency vs. black.  If the input sources aren't consuming the entire canvas I would certainly rather have a transparent background while taking a screenshot.

WizardCM commented:
Yeah I considered the same - a screenshot of a single source should support transparency, whereas whole canvas makes the most sense with a black background.

12345678kkmmxx commented:
@cg2121
